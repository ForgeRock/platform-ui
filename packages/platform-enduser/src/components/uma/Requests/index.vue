<!-- Copyright 2019 ForgeRock AS. All Rights Reserved

Use of this code requires a commercial software license with ForgeRock AS.
or with one of its affiliates. All use shall be exclusively subject
to such license between the licensee and ForgeRock AS. -->
<template>
  <div>
    <template>
      <div id="listView">
        <BCard
          no-body
          class="mt-4">
          <BListGroup flush>
            <BListGroupItem
              v-for="(request, index) in requests"
              :key="index">
              <div class="d-sm-flex">
                <div class="media-body align-self-center mb-2 mb-sm-0">
                  <div class="mb-2">
                    <small><strong>{{ request.user }}</strong> {{ $t('pages.uma.requests.requestedAccess') }}</small>
                  </div>
                  <div class="media mb-1">
                    <div class="d-flex mr-2 align-self-center">
                      <FrFallbackImage
                        v-if="request.icon_uri"
                        :src="request.icon_uri"
                        fallback="description" />
                    </div>
                    <div class="media-body align-self-center">
                      <div class="media-body align-self-center">
                        {{ request.resource }}
                      </div>
                    </div>
                  </div>
                  <small class="text-muted">
                    {{ request.when | formatTime }}
                  </small>
                </div>
                <div
                  class="d-flex justify-content-start ml-sm-3 align-self-center"
                  v-if="!request.decision">
                  <a
                    href="#"
                    class="pr-3"
                    @click="finalizeAccess(request, index, 'approve')">
                    {{ $t('pages.uma.requests.allow') }}
                  </a>
                  <a
                    href="#"
                    class="px-2"
                    @click="finalizeAccess(request, index, 'deny')">
                    {{ $t('pages.uma.requests.deny') }}
                  </a>
                </div>
                <div
                  class="d-flex justify-content-start ml-sm-3 align-self-center"
                  v-if="request.decision">
                  <div
                    class="allow text-success"
                    v-if="request.allowed">
                    <i class="material-icons-outlined font-weight-bolder md-16 mb-1">
                      check
                    </i>
                    {{ $t('pages.uma.requests.allowed') }}
                  </div>
                  <div
                    class="deny text-danger"
                    v-if="!request.allowed">
                    <i class="material-icons-outlined font-weight-bolder md-16 mb-1">
                      block
                    </i>
                    {{ $t('pages.uma.requests.denied') }}
                  </div>
                </div>
              </div>
            </BListGroupItem>
          </BListGroup>
        </BCard>
      </div>
    </template>
  </div>
</template>

<script>
import moment from 'moment';
import FallbackImage from '@/components/utils/FallbackImage';

/**
 * @description Allows user to request access to a resource
 * */
export default {
  name: 'Requests',
  components: {
    FrFallbackImage: FallbackImage,
  },
  props: {
    requests: {
      type: Array,
      default: () => [],
    },
  },
  filters: {
    formatTime(dateString) {
      const eventDate = moment(dateString);

      if (eventDate.isSame(moment(), 'day')) {
        return eventDate.fromNow();
      }
      return eventDate.format('LT');
    },
  },
  methods: {
    finalizeAccess(request, index, action) {
      this.requests[index].decision = true;

      // eslint-disable-next-line no-underscore-dangle
      this.$emit('finalizeResourceAccess', request._id, action, {
        scopes: request.permissions,
        onSuccess: () => {
          if (action === 'approve') {
            this.requests[index].allowed = true;
          }
        },
      });
    },
  },
};
</script>

<style lang="scss" scoped>
</style>
