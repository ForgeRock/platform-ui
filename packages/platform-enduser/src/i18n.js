/**
 * Copyright (c) 2020-2024 ForgeRock. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */

import { createI18n } from 'vue-i18n';
import { merge } from 'lodash';
import dateTimeFormats from '@forgerock/platform-shared/src/dateTimeFormats.json';

function loadLocaleMessages() {
  const locales = require.context('./locales', true, /[A-Za-z0-9-_,\s]+\.json$/i);
  // Can't use workspace symlink due to jest issues
  const sharedLocales = require.context('../../platform-shared/src/locales', true, /[A-Za-z0-9-_,\s]+\.json$/i);
  const messages = {};

  locales.keys().forEach((key) => {
    const matched = key.match(/([A-Za-z0-9-_]+)\./i);
    if (matched && matched.length > 1) {
      const locale = matched[1];
      let generatedTranslation = locales(key);

      // Check for shared local and merge in if it exists
      const sharedLocaleExists = sharedLocales.keys().includes(key);
      if (sharedLocaleExists) {
        generatedTranslation = merge(sharedLocales(key), generatedTranslation);
      }

      messages[locale] = generatedTranslation;
    }
  });

  return messages;
}

export default createI18n({
  messages: loadLocaleMessages(),
  silentFallbackWarn: true,
  silentTranslationWarn: true,
  dateTimeFormats,
});
