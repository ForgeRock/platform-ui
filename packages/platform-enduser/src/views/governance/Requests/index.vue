<!-- Copyright (c) 2023 ForgeRock. All rights reserved.

This software may be modified and distributed under the terms
of the MIT license. See the LICENSE file for details. -->
<template>
  <BContainer>
    <div class="mt-5">
      <FrHeader
        :title="$t('pageTitles.MyRequests')"
        :subtitle="$t('governance.accessRequests.myRequests.subtitle')" />
      <BCard
        no-body
        class="card-tabs-vertical" />
    </div>
  </BContainer>
</template>

<script>
import {
  BCard,
  BContainer,
} from 'bootstrap-vue';
import FrHeader from '@forgerock/platform-shared/src/components/PageHeader';
/**
 * @description Landing page for Requests
 */
export default {
  name: 'Requests',
  components: {
    BCard,
    BContainer,
    FrHeader,
  },
  data() {
    return {
    };
  },
};
</script>
