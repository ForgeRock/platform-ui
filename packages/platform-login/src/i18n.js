/**
 * Copyright (c) 2020-2021 ForgeRock. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */

import Vue from 'vue';
import VueI18n from 'vue-i18n';
import { merge } from 'lodash';
import dateTimeFormats from '@forgerock/platform-shared/src/dateTimeFormats.json';

Vue.use(VueI18n);

function loadLocaleMessages() {
  const locales = require.context('./locales', true, /[A-Za-z0-9-_,\s]+\.json$/i);
  // Can't use workspace symlink due to jest issues
  const sharedLocales = require.context('../../platform-shared/src/locales', true, /[A-Za-z0-9-_,\s]+\.json$/i);
  const messages = {};

  locales.keys().forEach((key) => {
    const matched = key.match(/([A-Za-z0-9-_]+)\./i);
    if (matched && matched.length > 1) {
      const locale = matched[1];
      let generatedTranslation = locales(key);

      // Check for shared local and merge in if it exists
      const sharedLocaleExists = sharedLocales.keys().includes(key);
      if (sharedLocaleExists) {
        generatedTranslation = merge(sharedLocales(key), generatedTranslation);
      }

      messages[locale] = generatedTranslation;
    }
  });

  return messages;
}

Vue.mixin({
  methods: {
    translationExists(path) {
      return this.$t(path) !== path;
    },
  },
});

export default new VueI18n({
  locale: process.env.VUE_APP_I18N_LOCALE || 'en',
  fallbackLocale: process.env.VUE_APP_I18N_FALLBACK_LOCALE || 'en',
  messages: loadLocaleMessages(),
  silentTranslationWarn: true,
  dateTimeFormats,
});
