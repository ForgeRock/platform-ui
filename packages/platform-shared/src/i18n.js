/**
 * Copyright 2019 ForgeRock AS. All Rights Reserved
 *
 * Use of this code requires a commercial software license with ForgeRock AS.
 * or with one of its affiliates. All use shall be exclusively subject
 * to such license between the licensee and ForgeRock AS.
 */
import Vue from 'vue';
import VueI18n from 'vue-i18n';
import dateTimeFormats from './dateTimeFormats.json';

Vue.use(VueI18n);

function loadLocaleMessages() {
  const locales = require.context('./locales', true, /[A-Za-z0-9-_,\s]+\.json$/i);
  const messages = {};

  locales.keys().forEach((key) => {
    const matched = key.match(/([A-Za-z0-9-_]+)\./i);
    if (matched && matched.length > 1) {
      const locale = matched[1];
      messages[locale] = locales(key);
    }
  });

  return messages;
}

export default new VueI18n({
  locale: process.env.VUE_APP_I18N_LOCALE || 'en',
  fallbackLocale: process.env.VUE_APP_I18N_FALLBACK_LOCALE || 'en',
  messages: loadLocaleMessages(),
  silentTranslationWarn: true,
  dateTimeFormats,
});
