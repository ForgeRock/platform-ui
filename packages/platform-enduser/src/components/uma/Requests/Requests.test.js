import BootstrapVue from 'bootstrap-vue';
import { createLocalVue, shallowMount } from '@vue/test-utils';
import i18n from '@/i18n';
import Requests from '@/components/uma/Requests';

const localVue = createLocalVue();
localVue.use(BootstrapVue);

Requests.components['fr-fallback-image'] = jest.fn();

const requests = [
  {
    _id: 'be4cb21e-74a1-4498-aabd-310b6b5bb3415',
    user: 'phil',
    resource: 'Ultrasound',
    when: 1529937902420,
    allowed: false,
    decision: false,
    permissions: [
      'download',
    ],
  },
  {
    _id: '7de54a3b-1446-4ada-a8a5-3cd00739180613',
    user: 'steve',
    resource: 'Ultrasound',
    when: 1529937934625,
    allowed: false,
    decision: false,
    permissions: [
      'download',
    ],
  },
];

describe('UMA Requests Component', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallowMount(Requests, {
      localVue,
      i18n,
      propsData: {
        requests,
      },
    });
  });

  afterEach(() => {
    wrapper = undefined;
  });

  it('Requests page loaded', () => {
    expect(wrapper.name()).toBe('Requests');
    wrapper = shallowMount(Requests, {
      localVue,
      i18n,
      filters: {
        formatTime() {
          return '7:45';
        }
      },
      propsData: {
        requests,
      },
    });
    expect(wrapper).toMatchSnapshot();
  });

  it('should format as relative time difference for events that occured today', () => {
    const eventToday = new Date();
    const offset = eventToday.getHours() - 1;

    expect(wrapper.vm.$options.filters.formatTime(eventToday.setHours(offset))).toBe('an hour ago');
  });

  it('should use actual time for events on previous days', () => {
    const eventDifferentDay = new Date();
    eventDifferentDay.setDate(eventDifferentDay.getDate() - 2);
    const formattedTime = wrapper.vm.$options.filters.formatTime(eventDifferentDay);

    expect(formattedTime.match(/\d{1,2}:\d{1,2} [AP]M/)).toBeTruthy();
  });

  it('should emit "finalizeResourceAccess" event', () => {
    wrapper.vm.finalizeAccess('12345', 0, 'approve');

    expect(wrapper.emitted('finalizeResourceAccess').length).toBe(1);
  });
});
