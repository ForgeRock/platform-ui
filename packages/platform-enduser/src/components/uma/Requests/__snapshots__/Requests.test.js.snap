// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UMA Requests Component Requests page loaded 1`] = `
<div>
  <div id="listView">
    <bcard-stub bodytag="div" title="" titletag="h4" subtitle="" subtitletag="h6" subtitletextvariant="muted" headertag="div" footertag="div" tag="div" nobody="true" class="mt-4">
      <blistgroup-stub tag="div" flush="true">
        <blistgroupitem-stub tag="div" target="_self" event="click" routertag="a">
          <div class="d-sm-flex">
            <div class="media-body align-self-center mb-2 mb-sm-0">
              <div class="mb-2"><small><strong>phil</strong> requested access to:</small></div>
              <div class="media mb-1">
                <div class="d-flex mr-2 align-self-center">
                  <!---->
                </div>
                <div class="media-body align-self-center">
                  <div class="media-body align-self-center">
                    Ultrasound
                  </div>
                </div>
              </div> <small class="text-muted">
                7:45
              </small>
            </div>
            <div class="d-flex justify-content-start ml-sm-3 align-self-center"><a href="#" class="pr-3">
                Allow
              </a> <a href="#" class="px-2">
                Deny
              </a></div>
            <!---->
          </div>
        </blistgroupitem-stub>
        <blistgroupitem-stub tag="div" target="_self" event="click" routertag="a">
          <div class="d-sm-flex">
            <div class="media-body align-self-center mb-2 mb-sm-0">
              <div class="mb-2"><small><strong>steve</strong> requested access to:</small></div>
              <div class="media mb-1">
                <div class="d-flex mr-2 align-self-center">
                  <!---->
                </div>
                <div class="media-body align-self-center">
                  <div class="media-body align-self-center">
                    Ultrasound
                  </div>
                </div>
              </div> <small class="text-muted">
                7:45
              </small>
            </div>
            <div class="d-flex justify-content-start ml-sm-3 align-self-center"><a href="#" class="pr-3">
                Allow
              </a> <a href="#" class="px-2">
                Deny
              </a></div>
            <!---->
          </div>
        </blistgroupitem-stub>
      </blistgroup-stub>
    </bcard-stub>
  </div>
</div>
`;
